New Metric page (`/metric/query`)

kamilmielnik opened the pull request:
Closes #37353

----


https://github.com/metabase/metabase/assets/6830683/7c98d89e-4c26-4529-97e7-0482bed8e994



### How to verify

1. "+ New" > "Metric"
2. Choose any data source
3. Choose any query shape (enforcing correct Metric query shape will be done in #37355)
4. Visualize the query - URL should still be `/metric/*`
5. Save - you should be able to provide description & choose any parent Collection
    - :warning:  Make sure to include the word "metric" in the metric name so that our mocks kick in
6. You should be redirected to `/metric/:slug` page (this page does not exist yet, it's being added in #37368)

nemanjaglumac committed:
Add new `metric` item to the menu

nemanjaglumac committed:
Add dummy route `/metric/query`

nemanjaglumac committed:
Open new metric in a notebook mode

nemanjaglumac committed:
Initialize new metric URL programatically using the existing helper

kamilmielnik committed:
Use type attribute in Question.prototype._serializeForUrl

kamilmielnik committed:
Extract getMultiStepTitle and add support for metrics

kamilmielnik committed:
Add support for metrics in getSingleStepTitle

kamilmielnik committed:
Combine getSingleStepTitle & getMultiStepTitle into getLabels

kamilmielnik committed:
Move nameInputPlaceholder computation into getLabels

kamilmielnik committed:
Use card.type in Urls.question

kamilmielnik committed:
Update getIsSavedQuestionChanged to account for metrics

kamilmielnik committed:
Remove dead code branch

kamilmielnik committed:
Always pass cardType to Urls.newQuestion to prevent Question.prototype.type from returning undefined

kamilmielnik committed:
Default type to "question" in getLabels for extra-safety

kamilmielnik committed:
Reorder conditions

kamilmielnik committed:
Use nullish coalescing operator instead of a ternary expression

kamilmielnik committed:
Remove no-longer-needed fallback

kamilmielnik commented on frontend/src/metabase/lib/urls/questions.ts:
```diff
@@ -105,9 +107,7 @@ export function newQuestion({
-
-  const entity = question.isDataset() ? "model" : "question";
-  const type = question.type() ?? entity;
+  const type = question.type();
```
The fallback is no longer needed thanks to 98b8579223391532c87276f89acbf271012c8c5f

kamilmielnik committed:
Set reasonable defaults in Question.create

kamilmielnik committed:
Use model's UI for metrics

kamilmielnik committed:
Make Urls.newQuestion safer

kamilmielnik committed:
Do not show app bar in metrics pages

kamilmielnik committed:
Disable metadata tab for metrics

kamilmielnik committed:
Hide tabs for metrics

kamilmielnik committed:
Merge branch 'metrics-v2' into 37353-new-metric-page

kamilmielnik committed:
Merge branch 'metrics-v2' into 37353-new-metric-page

kamilmielnik committed:
Rename dataset to question

kamilmielnik committed:
Rename dataset to question

kamilmielnik committed:
Account for metric-based ad-hoc questions in updateQuestion

kamilmielnik committed:
Account for metrics in View

kamilmielnik committed:
Migrate Question.prototype.isDataset in hasQuestionCacheSection

kamilmielnik committed:
Migrate Question.prototype.isDataset in metabase-enterprise/moderation

kamilmielnik committed:
Migrate Question.prototype.isDataset in ModelDetailPage

kamilmielnik committed:
Migrate Question.prototype.isDataset in navigateToNewCardFromDashboard

kamilmielnik committed:
Migrate Question.prototype.isDataset in handleQBInit

kamilmielnik committed:
Migrate Question.prototype.isDataset in Visualization

kamilmielnik committed:
Migrate Question.prototype.isDataset in QueryModals

kamilmielnik committed:
Migrate Question.prototype.isDataset in QuestionPane

kamilmielnik committed:
Migrate Question.prototype.isDataset in ObjectDetailView

kamilmielnik committed:
Migrate Question.prototype.isDataset in ArchiveQuestionModal

kamilmielnik committed:
Migrate Question.prototype.isDataset in Question.prototype.composeDataset

kamilmielnik committed:
Migrate Question.prototype.isDataset in getNativeQueryTable

kamilmielnik committed:
Migrate Question.prototype.isDataset in ViewFooter

kamilmielnik committed:
Convert SavedQuestionIntroModal to TS

kamilmielnik committed:
Migrate Question.prototype.isDataset in SavedQuestionIntroModal

kamilmielnik committed:
Migrate Question.prototype.isDataset in QuestionMoveToast

kamilmielnik committed:
Migrate Question.prototype.isDataset in QuestionInfoSidebar

kamilmielnik committed:
Fix QuestionMoveToast usage

kamilmielnik committed:
Fix unit tests

kamilmielnik committed:
Fix unit tests

nemanjaglumac commented on frontend/src/metabase/components/NewItemMenu/NewItemMenu.tsx:
```diff
@@ -136,9 +138,10 @@ const NewItemMenu = ({
-          mode: "notebook",
-          collectionId,
+          mode: "query",
+          creationType: "custom_question",
```
Let's remove this until it's needed for analytics.

kamilmielnik replied:
d07a15c1e79b1362757af80cb222064938014aa7

kamilmielnik committed:
Make e2e test less flaky

kamilmielnik committed:
Make code work exactly like before

kamilmielnik committed:
Revert breaking change

kamilmielnik committed:
Merge branch 'migrate-is-dataset' into 37353-new-metric-page

kamilmielnik committed:
Migrate Question.prototype.isDataset in QueryModals

kamilmielnik committed:
Merge branch 'metrics-v2' into 37353-new-metric-page

kamilmielnik committed:
Add routing to state typing

kamilmielnik committed:
Add getRouting selector

kamilmielnik committed:
Infer entity type from URL when converting question back

kamilmielnik committed:
Remove creationType for metrics

kamilmielnik committed:
Make getCardTypeFromLocation more strict

kamilmielnik commented on frontend/src/metabase/query_builder/actions/core/updateQuestion.ts:
```diff
@@ -129,17 +131,19 @@ export const updateQuestion = (
     // this `updateQuestion` action, so they end up triggering the action with the altered question.
     if (queryBuilderMode === "dataset" && !newQuestion.isDataset()) {
-      newQuestion = newQuestion.setType("model");
+      const routing = getRouting(getState());
+      const type = getCardTypeFromLocation(routing.locationBeforeTransitions);
+      newQuestion = newQuestion.setType(type);
```
This is an extension to the existing "hack".
Now `queryBuilderMode === "dataset"` is shared by both models and metrics.
URL is the only place where I could find information on what type of entity are we dealing with.

Alternative solutions I considered:
- remove the original hack - [this is the best option, but also the most difficult](https://metaboat.slack.com/archives/C505ZNNH4/p1707400291390129)
- add one more piece of information to `state.qb.uiControls` indicating what kind of entity did we start with - simple, but might have edge cases
- introduce a new `queryBuilderMode` - would require a lot of changes

kamilmielnik replied:
I managed to get rid of this: c1d3db383613b4bdebfdbfcfb1620b144d3e1503

kamilmielnik committed:
Replace Question.prototype.isDataset usage with Question.prototype.type

kamilmielnik committed:
Add mocks for routing state

kamilmielnik committed:
Use Location in getCardTypeFromLocation

kamilmielnik committed:
Merge branch 'metrics-v2' into 37353-new-metric-page

kamilmielnik committed:
Ensure initialState contains "routing" only when using routerReducer
- this avoids getting warning from redux saying that initial state contains an extraneous key

kamilmielnik committed:
Extract createMockLocation

kamilmielnik committed:
Remove getCardTypeFromLocation hack, use state instead

kamilmielnik committed:
Revert routing changes

kamilmielnik committed:
Reduce diff

nemanjaglumac commented on frontend/src/metabase/components/NewItemMenu/NewItemMenu.tsx:
```diff
@@ -136,9 +138,9 @@ const NewItemMenu = ({
         link: Urls.newQuestion({
-          mode: "notebook",
-          collectionId,
+          mode: "query",
```
I don't understand why did we change this to `query`? If you take a look at the question flow, the URL says `/notebook` which seems to reflect the UI state correctly.

Now, compare it to the metric:
![image](https://github.com/metabase/metabase/assets/31325167/d0b2f13b-d1d8-447b-9c0a-c8ef29c618ce)

Unless there was a really good reason for this change, I think we should change it back to the "notebook".

kamilmielnik replied:
> If you take a look at the question flow

We want to follow the model flow.

nemanjaglumac replied:
Ok, I was about to correct myself because I've just opened a model and saw that it has a `/query` in the URL.

Admittedly, this is adding new layer of confusion but that has nothing to do with your changes here.
Ignore my comment.

nemanjaglumac commented on frontend/src/metabase/containers/SaveQuestionModal.tsx:
```diff
@@ -170,27 +190,18 @@ export const SaveQuestionModal = ({
 
-  const singleStepTitle = getSingleStepTitle(questionType, showSaveType);
-
+  const { multiStepTitle, singleStepTitle, nameInputPlaceholder } = getLabels(
+    question,
+    showSaveType,
+  );
```
So much cleaner 👏

nemanjaglumac commented on frontend/src/metabase/query_builder/components/DatasetEditor/DatasetEditor.jsx:
```diff
@@ -444,21 +444,26 @@ function DatasetEditor(props) {
-              },
-            ]}
-          />
+          // Metadata tab is temporarily disabled for metrics.
+          // It should be enabled in #37993
+          // @see https://github.com/metabase/metabase/issues/37993
+          question.type() === "metric" ? null : (
+            <EditorTabs
+              currentTab={datasetEditorTab}
+              onChange={onChangeEditorTab}
+              options={[
+                { id: "query", name: t`Query`, icon: "notebook" },
+                {
+                  id: "metadata",
+                  name: t`Metadata`,
+                  icon: "label",
+                  disabled: !resultsMetadata,
+                },
+              ]}
+            />
+          )
```
👌

nemanjaglumac commented on frontend/src/metabase/query_builder/components/view/View.jsx:
```diff
@@ -350,7 +350,10 @@ class View extends Component {
     }
 
-    if (question.isDataset() && queryBuilderMode === "dataset") {
+    const isModelOrMetric =
+      question.type() === "model" || question.type() === "metric";
```
Did you intentionally avoid using `question.type() !== "question"` for the sake of being more explicit?

kamilmielnik replied:
Also because `question.type()` can be `undefined`

nemanjaglumac replied:
I thought that we're falling back to "question" if it's undefined.
But yeah, until we have the type safety here, this makes sense.

Thank you

nemanjaglumac reviewed (approved):
Very clean.
Let's merge this. 🚢

kamilmielnik commented:
> Very clean. Let's merge this. 🚢

Thanks for review @nemanjaglumac :pray: 
Some unit tests are still failing, I need to sort it out first.
If changes turn out to be large I'll request a re-review.

nemanjaglumac commented on frontend/src/metabase/query_builder/actions/core/updateQuestion.ts:
```diff
@@ -129,17 +130,23 @@ export const updateQuestion = (
     }
 
-    // This scenario happens because the DatasetQueryEditor converts the dataset/model question into a normal question
+    // This scenario happens because the DatasetQueryEditor converts the model/metric question into a normal question
```
```suggestion
    // This scenario happens because the DatasetQueryEditor converts a model or a metric into a normal question
```

The previous change could be easily misinterpreted.

kamilmielnik committed:
Update updateQuestion unit tests

kamilmielnik committed:
Update QuestionDataSource unit tests

kamilmielnik committed:
Update QueryBuilder.unsaved-changes-warning.unit.spec.tsx

kamilmielnik committed:
Update ActionCreatorModal unit tests

kamilmielnik committed:
Update SaveQuestionModal unit tests

kamilmielnik committed:
Update comment

Co-authored-by: Nemanja Glumac <31325167+nemanjaglumac@users.noreply.github.com>

kamilmielnik committed:
Remove question type toggling
- it should no longer be required
- see @https://metaboat.slack.com/archives/C505ZNNH4/p1707400291390129

kamilmielnik committed:
Remove outdated comment

kamilmielnik committed:
Update unit tests

kamilmielnik committed:
Merge branch 'remove-question-type-toggling' into 37353-new-metric-page

kamilmielnik committed:
Update reference/utils unit tests

kamilmielnik committed:
Revert unnecessary changes

kamilmielnik committed:
Merge branch 'metrics-v2' into 37353-new-metric-page

kamilmielnik committed:
Fix unit tests
